File: ChromeDriverTest.js
'use strict';

const webdriver = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');

function usage() {
  console.log('usage: "' + process.argv[0]
      + '" "' + process.argv[1] + '" [optins...]');
  console.log('\t-driver driverPath');
  console.log('\t-chrome chromePath');
  console.log('\t-data userDataPath');
  console.log('\t-service (use ChromeDriver service)');
  console.log('\t-remote http://ipAddress:port (connect to existing ChromeDriver)');
  console.log('\t-url URL');
  console.log('\t-log logFilePath');
  console.log('\t-save htmlFilePath (save page contents to file)');
  console.log('\t-name (display browser name)');
  console.log('\t-title (display page title)');
  console.log('\t-cookies (cookies demo)');
  console.log('\t-position (set position demo)');
  console.log('\t-size (set size demo)');
  console.log('\t-keep (do not quit ChromeDriver)');
  process.exit(1);
}

let driverPath = undefined;
let chromePath = undefined;
let userDataDir = undefined;
let useService = false;
let remotePort = undefined;
let url = 'https://www.google.com/';
let logPath = undefined;
let html = undefined;
let remoteDebuggingPort = undefined;
let name = false;
let title = false;
let testCookies = false;
let testPosition = false;
let testSize = false;
let quit = true;

// process.argv[0] is path to node, process.argv[1] is path to JS file.
// User arguments start with process.argv[2].
for (let argi = 2; argi < process.argv.length; argi++) {
  if (process.argv[argi][0] !== '-') usage();

  if (process.argv[argi] === '-driver' && argi + 1 < process.argv.length)
    driverPath = process.argv[++argi];
  else if (process.argv[argi] === '-chrome' && argi + 1 < process.argv.length)
    chromePath = process.argv[++argi];
  else if (process.argv[argi] === '-data' && argi + 1 < process.argv.length)
    userDataDir = process.argv[++argi];
  else if (process.argv[argi] === '-service')
    useService = true;
  else if (process.argv[argi] === '-remote' && argi + 1 < process.argv.length)
    remotePort = process.argv[++argi];
  else if (process.argv[argi] === '-url' && argi + 1 < process.argv.length)
    url = process.argv[++argi];
  else if (process.argv[argi] === '-log' && argi + 1 < process.argv.length)
    logPath = process.argv[++argi];
  else if (process.argv[argi] === '-save' && argi + 1 < process.argv.length)
    html = process.argv[++argi];
  else if (process.argv[argi] === '-debug' && argi + 1 < process.argv.length)
    remoteDebuggingPort = process.argv[++argi];
  else if (process.argv[argi] === '-name')
    name = true;
  else if (process.argv[argi] === '-title')
    title = true;
  else if (process.argv[argi] === '-cookies')
    testCookies = true;
  else if (process.argv[argi] === '-position')
    testPosition = true;
  else if (process.argv[argi] === '-size')
    testSize = true;
  else if (process.argv[argi] === '-keep')
    quit = false;
  else
    usage();
}

let driver;

if (remotePort !== undefined) {
  driver = new webdriver.Builder()
      .forBrowser('chrome')
      .usingServer(remotePort)
      .build();
} else {
  let builder = new chrome.ServiceBuilder(driverPath);
  const options = new chrome.Options();

  if (chromePath !== undefined)
    options.setChromeBinaryPath(chromePath);

  if (userDataDir !== undefined)
    options.addArguments('user-data-dir=' + userDataDir);

  if (logPath !== undefined)
    builder = builder.loggingTo(logPath).enableVerboseLogging();

  const service = builder.build();
  driver = chrome.Driver.createSession(options, service);
}

let promise = Promise.resolve();

if (name) {
  promise = promise.then(() => {
    driver.getCapabilities().then(capabilities => {
      console.log('Browser name: '
          + capabilities.get(webdriver.Capability.BROWSER_NAME));
    });
  });
}

promise = promise.then(() => driver.get(url));

if (title) {
  promise = promise.then(() => {
    driver.getTitle().then(title => console.log('Page title: ' + title));
  });
  promise = promise.then(() => {
    driver.getCurrentUrl().then(url => console.log('Page URL: ' + url));
  });
}

if (quit) {
  promise = promise
      .then(() => new Promise(resolve => setTimeout(resolve, 3000)))
      .then(() => driver.quit());
}

File: ChromeDriverTest.java
import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Set;
import org.openqa.selenium.Cookie;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.Point;
import org.openqa.selenium.Proxy;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

public class ChromeDriverTest {
  static void usage() {
    System.err.println("Available options:");
    System.err.println("\t-driver driverPath");
    System.err.println("\t-chrome chromePath");
    System.err.println("\t-data userDataPath");
    System.err.println("\t-service (use ChromeDriver service)");
    System.err.println("\t-remote http://ipAddress:port (connect to existing ChromeDriver)");
    System.err.println("\t-url URL");
    System.err.println("\t-log logFilePath");
    System.err.println("\t-save htmlFilePath (save page contents to file)");
    System.err.println("\t-debug ipAddress:port");
    System.err.println("\t-name (display browser name)");
    System.err.println("\t-title (display page title)");
    System.err.println("\t-cookies (cookies demo)");
    System.err.println("\t-position (set position demo)");
    System.err.println("\t-size (set size demo)");
    System.err.println("\t-proxy proxyServer:port");
    System.err.println("\t-keep (do not quit ChromeDriver)");
    System.exit(1);
  }

  public static void main(String[] args) throws Exception {
    String driverPath = null;
    boolean useService = false;
    URL remotePort = null;
    ChromeOptions chromeOptions = new ChromeOptions();
    String logPath = null;
    String url = "http://www.google.com/";
    String htmlPath = null;
    boolean name = false;
    boolean title = false;
    boolean quit = true;
    boolean testCookies = false;
    boolean testPosition = false;
    boolean testSize = false;

    for (int argi = 0; argi < args.length; argi++) {
      if (!args[argi].startsWith("-"))
        usage();

      if (args[argi].equals("-driver") && argi + 1 < args.length)
        driverPath = args[++argi];
      else if (args[argi].equals("-chrome") && argi + 1 < args.length)
        chromeOptions.setBinary(args[++argi]);
      else if (args[argi].equals("-data") && argi + 1 < args.length)
        chromeOptions.addArguments("user-data-dir=" + args[++argi]);
      else if (args[argi].equals("-service"))
        useService = true;
      else if (args[argi].equals("-remote") && argi + 1 < args.length)
        remotePort = new URL(args[++argi]);
      else if (args[argi].equals("-url") && argi + 1 < args.length)
        url = args[++argi];
      else if (args[argi].equals("-log") && argi + 1 < args.length)
        logPath = args[++argi];
      else if (args[argi].equals("-save") && argi + 1 < args.length)
        htmlPath = args[++argi];
      else if (args[argi].equals("-debug") && argi + 1 < args.length)
        // Chrome should have been started with --remote-debugging-port=port
        chromeOptions.setExperimentalOption("debuggerAddress", args[++argi]);
      else if (args[argi].equals("-name"))
        name = true;
      else if (args[argi].equals("-title"))
        title = true;
      else if (args[argi].equals("-cookies"))
        testCookies = true;
      else if (args[argi].equals("-position"))
        testPosition = true;
      else if (args[argi].equals("-size"))
        testSize = true;
      else if (args[argi].equals("-proxy") && argi + 1 < args.length) {
        Proxy proxy = new Proxy();
        proxy.setHttpProxy(args[++argi]);
        chromeOptions.setCapability("proxy", proxy);
      }
      else if (args[argi].equals("-keep"))
        quit = false;
      else
        usage();
    }

    if (useService && remotePort != null)
    {
      System.err.println("Can't use both -service and -remote options");
      System.exit(1);
    }

    ChromeDriverService service = null;
    if (remotePort == null) {
      ChromeDriverService.Builder builder = new ChromeDriverService.Builder();
      if (driverPath != null)
        builder
            .usingDriverExecutable(new File(driverPath));
      if (logPath != null)
        builder
            .withLogFile(new File(logPath))
            .withVerbose(true);
      service = builder.build();
    }

    if (useService) {
      service.start();
      remotePort = service.getUrl();
      System.out.println("Service URL: " + remotePort);
    }

    RemoteWebDriver driver;

    if (remotePort != null)
      driver = new RemoteWebDriver(remotePort, chromeOptions);
    else {
      driver = new ChromeDriver(service, chromeOptions);
      /* Alternative code. The following code is probably more commonly seen.
       * However, it is not multi-thread safe, as System.setProperty affects
       * the entire process.
      if (driverPath != null)
        System.setProperty("webdriver.chrome.driver", driverPath);
      if (logPath != null) {
        System.setProperty("webdriver.chrome.logfile", logPath);
        System.setProperty("webdriver.chrome.verboseLogging", "true");
      }
      driver = new ChromeDriver(chromeOptions);
      */
    }

    if (name)
      System.out.println("Browser name: " + driver.getCapabilities().getBrowserName());

    driver.get(url);

    if (title)
    {
      System.out.println("Page title: " + driver.getTitle());
      System.out.println("Page URL: " + driver.getCurrentUrl());
    }

    if (htmlPath != null)
    {
      OutputStreamWriter out = new OutputStreamWriter(
          new FileOutputStream(htmlPath), StandardCharsets.UTF_8);
      out.write(driver.getPageSource());
    }

    if (testCookies)
    {
      driver.manage().addCookie(new Cookie("foo", "bar", "/"));
      System.out.println("Cookies:");
      for (Cookie cookie: driver.manage().getCookies())
        System.out.println("\t" + cookie);
    }

    if (testPosition)
    {
        WebDriver.Window window = driver.manage().window();
        Point position = window.getPosition();
        int x = position.x;
        int y = position.y;
        System.out.printf("Current window position (%d, %d)\n", x, y);

        x += 100;
        y += 100;
        System.out.printf("Attempting to change window position to (%d, %d)\n", x, y);
        window.setPosition(new Point(x, y));

        position = window.getPosition();
        System.out.printf("New window position (%d, %d)\n", position.x, position.y);
    }

    if (testSize)
    {
        WebDriver.Window window = driver.manage().window();
        Dimension size = window.getSize();
        int width = size.width;
        int height = size.height;
        System.out.printf("Current window size %dx%d\n", width, height);

        width /= 2;
        height /= 2;
        System.out.printf("Attempting to change window size to %dx%d\n", width, height);
        window.setSize(new Dimension(width, height));

        size = window.getSize();
        System.out.printf("New window size %dx%d\n", size.width, size.height);
    }

    if (quit)
    {
      Thread.sleep(3000);
      driver.quit();
    }
  }
}
